### PR TITLE
Bind OID4VP auth completion to the originating flow

### PR DESCRIPTION
**Pull request by Ogenbertrand**
This PR enforces single ownership of the OID4VP authentication flow.
- For browser/OIDC flows, authentication completion remains bound to the originating OIDC authentication session.
- For API-only flows, the `authorization_code` is no longer exposed through the public status endpoint and must instead be redeemed through a verifier-bound `/code` endpoint.

closes: https://github.com/ADORSYS-GIS/digital-identity-arch-review-mab/issues/11

### CODE REVIEW

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 1]
```diff
@@ -85,7 +92,14 @@ public Response getAuthenticationRequest(@QueryParam(OAuth2Constants.CLIENT_ID)
...
         }
 
-        AuthorizationContext authContext = startAuthentication(clientId, null);
+        AuthorizationContext authContext;
+        try {
+            authContext = startAuthentication(clientId, null, codeChallenge, codeChallengeMethod);
+        } catch (IllegalArgumentException e) {
+            throw new BadRequestException(
+                    errorResponse(Response.Status.BAD_REQUEST, OAuthErrorException.INVALID_REQUEST, e.getMessage()), e);
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java:111]: Return a contextualized, generic error message instead of leaking `e.getMessage()` without proper control. The specific cause will naturally be found in logs because it is appended to the exception.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 2]
```diff
@@ -196,27 +210,106 @@ public Response getAuthorizationContextState(@PathParam("transactionId") String
...
+                    OAuthErrorException.INVALID_REQUEST,
+                    "Authorization has not completed successfully"));
+        }
+
+        if (!StringUtil.isBlank(authorizationContext.getParentAuthSessionId())) {
+            throw new BadRequestException(errorResponse(
+                    Response.Status.BAD_REQUEST,
+                    OAuthErrorException.INVALID_REQUEST,
+                    "Authorization code must be completed through the bound OIDC session"));
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: This check is excessive. It will be covered by the absence of a code challenge in the session storage.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 3]
```diff
@@ -262,6 +355,36 @@ private Response errorResponse(Response.Status status, String error, String erro
...
+        boolean hasCodeChallenge = !StringUtil.isBlank(codeChallenge);
+        boolean hasCodeChallengeMethod = !StringUtil.isBlank(codeChallengeMethod);
+
+        if (hasParentAuthSession) {
+            if (hasCodeChallenge || hasCodeChallengeMethod) {
+                throw new IllegalArgumentException(
+                        "OIDC-bound OpenID4VP flows do not accept code challenge parameters");
+            }
+            return null;
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: Do not bother adding this check. We control that flow.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 4]
```diff
@@ -262,6 +355,36 @@ private Response errorResponse(Response.Status status, String error, String erro
...
+
+        if (!hasCodeChallenge && !hasCodeChallengeMethod) {
+            throw new IllegalArgumentException(
+                    "Public API requests must provide a code challenge for authorization code redemption");
+        }
+
+        if (!hasCodeChallenge || !hasCodeChallengeMethod) {
+            throw new IllegalArgumentException("Both code_challenge and code_challenge_method are required");
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: I suggest you combine these checks and the error message. When appropriate, concise code is always preferable for maintainability.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java [hunk 5]
```diff
@@ -262,6 +355,36 @@ private Response errorResponse(Response.Status status, String error, String erro
...
+
+        if (!OAuth2Constants.PKCE_METHOD_S256.equals(codeChallengeMethod)) {
+            throw new IllegalArgumentException("Only S256 code challenge method is supported");
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: No strong opinion, but any reason why?

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java [hunk 6]
```diff
@@ -50,8 +50,11 @@ public abstract class OID4VPBaseUserAuthEndpointTest extends OID4VPBaseKeycloakT
      */
     protected String testSuccessfulAuthentication(String sdJwt, TestOpts opts) throws Exception {
         // Retrieve an authorization request
+        ApiFlowData apiFlow = opts.getAuthContext() == null ? startApiAuthorizationRequest() : null;
         AuthorizationContext authContext =
-                Optional.ofNullable(opts.getAuthContext()).orElseGet(this::requestAuthorizationRequest);
+                Optional.ofNullable(opts.getAuthContext()).orElseGet(() -> apiFlow.authContext());
+        String codeVerifier = Optional.ofNullable(opts.getCodeVerifier())
+                .orElseGet(() -> apiFlow == null ? null : apiFlow.codeVerifier());
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java]: There are too many repetitive checks here. Consider refactoring for improved readability.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java [hunk 7]
```diff
@@ -63,8 +66,13 @@ protected String testSuccessfulAuthentication(String sdJwt, TestOpts opts) throw
         AuthorizationContext statusPayload = parseAuthorizationContext(statusResponse);
         assertEquals(AuthorizationContextStatus.SUCCESS, statusPayload.getStatus());
 
-        // Exchange authorization code for access token
+        // Redeem authorization code when it is not disclosed in the status response
         String authCode = statusPayload.getAuthorizationCode();
+        if (authCode == null) {
+            assertNotNull(codeVerifier, "Code verifier should not be null for API flows");
+            authCode = redeemAuthorizationCode(authContext.getTransactionId(), codeVerifier);
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java]: Regarding the `assertNotNull`, what if we add a test that sends a null codeVerifier to test redeemAuthorizationCode?

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java [hunk 8]
```diff
@@ -184,6 +188,75 @@ public void shouldNotDiscloseStatusWithRequestIDs() throws Exception {
                 "Only transaction IDs should enable polling authorization statuses");
     }
 
+    @Test
+    public void shouldNotDiscloseAuthorizationCodeInStatus_ForApiFlows() throws Exception {
+        // Request a valid SD-JWT credential from Keycloak to use for authentication
+        String sdJwt = sdJwtVPTestUtils.requestSdJwtCredential(VCT_CONFIG_DEFAULT, TEST_USER);
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:198]: You could just add the target assertion in existing paths.
I guess it is this.

```java
        assertNull(statusPayload.getAuthorizationCode(), "authorization_code must stay hidden for API flows");
```

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java [hunk 9]
```diff
@@ -184,6 +188,75 @@ public void shouldNotDiscloseStatusWithRequestIDs() throws Exception {
...
+        // Polling should reveal status, but not the redeemable authorization code
+        HttpResponse statusResponse = fetchAuthenticationStatus(authContext.getTransactionId());
+        AuthorizationContext statusPayload = parseAuthorizationContext(statusResponse);
+        assertEquals(AuthorizationContextStatus.SUCCESS, statusPayload.getStatus());
+        assertNull(statusPayload.getAuthorizationCode(), "authorization_code must stay hidden for API flows");
+    }
+
+    @Test
+    public void shouldRedeemAuthorizationCodeWithValidVerifier_ForApiFlows() throws Exception {
+        // Request a valid SD-JWT credential from Keycloak to use for authentication
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java]: This is necessarily already covered by existing successful tests.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java [hunk 10]
```diff
@@ -184,6 +188,75 @@ public void shouldNotDiscloseStatusWithRequestIDs() throws Exception {
...
+                new BasicNameValuePair("transaction_id", authContext.getTransactionId()),
+                new BasicNameValuePair(OAuth2Constants.CODE_VERIFIER, "invalid-code-verifier"))));
+
+        HttpResponse codeResponse = httpClient.execute(httpPost);
+        assertEquals(HttpStatus.SC_BAD_REQUEST, codeResponse.getStatusLine().getStatusCode());
+
+        OAuth2ErrorRepresentation errorRep = parseErrorResponse(codeResponse);
+        assertEquals(OAuthErrorException.INVALID_GRANT, errorRep.getError());
+        assertEquals("Authorization code verifier not valid", errorRep.getErrorDescription());
+    }
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:230]: This test is needed but seems very verbose. Is it possible to reuse one of the existing testFailingAuthentication? Consider using the `testOpts.setCodeVerifier` you added in the PR.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java [hunk 11]
```diff
@@ -85,7 +85,11 @@ public AuthorizationRequestService(KeycloakSession session) {
      * Creates a fresh authorization request for user authentication.
      */
     public AuthorizationContext createAuthorizationRequest(
-            SdJwtAuthRequirements authReqs, AuthenticationSessionModel authSession, String parentAuthSessionId) {
+            SdJwtAuthRequirements authReqs,
+            AuthenticationSessionModel authSession,
+            String parentAuthSessionId,
+            String codeChallenge,
+            String codeChallengeMethod) {
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java]: I suggest you add a record to the code base, if not existing already, so these two variables can be passed around concisely.

**Review by IngridPuppet — COMMENTED** [review]
Great addition and implementation! Please check my minor comments below.

For now, I agree it is okay not to deal with OIDC-wrapped authentication, but I think we should create another ticket to tackle that next. Otherwise, it will still be possible to steal the authorization code (which can be freely converted into an access token) just with the transaction ID, in the same initial terms described by the issue associated with this PR.

It can be possible to generate PKCE details and forward the codeVerifier to the JavaScript in the login page.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 1 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java:111]: Thanks fixed

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 2 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: removed

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 3 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: removed

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 4 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: done

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java (hunk 6 above)
Comment (Ogenbertrand) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java]: fixed

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java (hunk 7 above)
Comment (Ogenbertrand) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java]: I added the test.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java (hunk 8 above)
Comment (Ogenbertrand) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:198]: yes fixed

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java (hunk 10 above)
Comment (Ogenbertrand) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:230]: fixed

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java (hunk 11 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java]: done, i added it as a method in this class, do you think we should have a separate class for it ?

**Comment by Ogenbertrand** [comment]
> Great addition and implementation! Please check my minor comments below.
> 
> For now, I agree it is okay not to deal with OIDC-wrapped authentication, but I think we should create another ticket to tackle that next. Otherwise, it will still be possible to steal the authorization code (which can be freely converted into an access token) just with the transaction ID, in the same initial terms described by the issue associated with this PR.
> 
> It can be possible to generate PKCE details and forward the codeVerifier to the JavaScript in the login page.

I'll createa follow up ticket to tackel authorization_code next.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java [hunk 12]
```diff
@@ -108,7 +122,7 @@ protected void testFailingAuthentication(
             throws Exception {
         // Retrieve an authorization request
         AuthorizationContext authContext =
-                Optional.ofNullable(opts.getAuthContext()).orElseGet(this::requestAuthorizationRequest);
+                opts.getAuthContext() == null ? requestAuthorizationRequest() : opts.getAuthContext();
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java]: How is this an improvment?

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java [hunk 13]
```diff
@@ -52,28 +52,20 @@ public abstract class OID4VPBaseUserAuthEndpointTest extends OID4VPBaseKeycloakT
      * Helper for successful flows.
      */
     protected String testSuccessfulAuthentication(String sdJwt, TestOpts opts) throws Exception {
-        // Retrieve an authorization request
-        ApiFlowData apiFlow = opts.getAuthContext() == null ? startApiAuthorizationRequest() : null;
-        AuthorizationContext authContext =
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java:56]: Keep the comment.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java [hunk 14]
```diff
@@ -154,6 +153,11 @@ public AuthorizationContext createAuthorizationRequest(
         return authorizationContext;
     }
 
+    /**
+     * Record to hold PKCE parameters for passing around concisely.
+     */
+    public record CodeChallengeDetails(String codeChallenge, String codeChallengeMethod) {}
+
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/service/AuthorizationRequestService.java]: Relocate this at the very end of the class.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java [hunk 15]
```diff
@@ -197,72 +193,35 @@ public void shouldNotDiscloseStatusWithRequestIDs() throws Exception {
...
-
-        // Complete the OID4VP exchange
-        HttpResponse response = sendAuthorizationResponse(sdJwt, requestObject, TestOpts.getDefault());
-        assertEquals(HttpStatus.SC_OK, response.getStatusLine().getStatusCode());
-
-        // Redeem the code explicitly with the matching verifier
-        String authCode = redeemAuthorizationCode(authContext.getTransactionId(), apiFlow.codeVerifier());
-        assertNotNull(authCode);
-        assertAuthenticatingUser(TestOpts.getDefault(), authCode);
+        TestOpts opts = TestOpts.getDefault().setAuthorizationContext(apiFlow.authContext());
```
Comment (IngridPuppet) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpointTest.java:205]: Add a comment to explain why you do this and why the test will fail because of this.

The test title says "withMissingVerifier". Clarify how this removes codeVerifier.

**Review by IngridPuppet — COMMENTED** [review]
I added more minor comments and reopened some conversations. Please could you take another look?

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java (hunk 13 above)
Comment (Ogenbertrand) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPBaseUserAuthEndpointTest.java:56]: I haven't deleted the comment. Can you check again ?

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java (hunk 5 above)
Comment (Ogenbertrand) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/OID4VPUserAuthEndpoint.java]: It's primarily best security practice and alignment with modern standards. no much reasons other than that.

**Review by IngridPuppet — APPROVED** [review]
LGTM. Merging to fast-forward the PR. I'll add minor updates in my PR next.
